src/rule_definition_tools.jl

```diff
@@ -330,53 +330,57 @@ macro non_differentiable(sig_expr)
+    @gensym kwargs
+    return @strip_linenos quote
+        function ($ChainRulesCore.frule)(_, $(map(esc, primal_sig_parts)...); $(esc(kwargs))...)
             $(__source__)
             # Julia functions always only have 1 output, so return a single DoesNotExist()
-            return ($(_with_kwargs_expr(primal_invoke)), DoesNotExist())
+            return ($(esc(_with_kwargs_expr(primal_invoke, kwargs))), $(DoesNotExist)())
         end
-    ))
+    end
 end
 
 function tuple_expression(primal_sig_parts)
     has_vararg = _isvararg(primal_sig_parts[end])
     return if !has_vararg
-        num_primal_inputs = length(primal_sig_parts) - 1 # - primal
-        Expr(:tuple, ntuple(_->DoesNotExist(), num_primal_inputs)...)
+        num_primal_inputs = length(primal_sig_parts)
+        Expr(:tuple, ntuple(_ -> :($(DoesNotExist)()), num_primal_inputs)...)
     else
-        num_primal_inputs = length(primal_sig_parts) - 2 # - primal and vararg
-        length_expr = :($(num_primal_inputs) + length($(_unconstrain(primal_sig_parts[end]))))
-        Expr(:call, :ntuple, Expr(:(->), :_, DoesNotExist()), length_expr)
+        num_primal_inputs = length(primal_sig_parts) - 1 # - vararg
+        length_expr = :($(num_primal_inputs) + $(length)($(esc(_unconstrain(primal_sig_parts[end])))))
+        Expr(:call, :ntuple, Expr(:(->), :($(esc(:_))), :($(DoesNotExist)())), length_expr)
     end
 end
 
 function _nondiff_rrule_expr(__source__, primal_sig_parts, primal_invoke)
+    esc_primal_sig_parts = map(esc, primal_sig_parts)
     tup_expr = tuple_expression(primal_sig_parts)
     primal_name = first(primal_invoke.args)
-    pullback_expr = Expr(
-        :function,
-        Expr(:call, propagator_name(primal_name, :pullback), :_),
-        Expr(:tuple, DoesNotExist(), Expr(:(...), tup_expr))
-    )
-    return esc(@strip_linenos quote
+    pullback_expr = @strip_linenos quote
+        function $(esc(propagator_name(primal_name, :pullback)))($(esc(:_)))
+            return $(tup_expr)
+        end
+    end
+
+    @gensym kwargs
+    return @strip_linenos quote
         # Manually defined kw version to save compiler work. See explanation in rules.jl
-        function (::Core.kwftype(typeof(ChainRulesCore.rrule)))(kwargs::Any, rrule::typeof(ChainRulesCore.rrule), $(primal_sig_parts...))
-            return ($(_with_kwargs_expr(primal_invoke)), $pullback_expr)
+        function (::$(Core.kwftype)($(typeof)($(rrule))))($(esc(kwargs))::$(Any), ::$(typeof)($(rrule)), $(esc_primal_sig_parts...))
+            return ($(esc(_with_kwargs_expr(primal_invoke, kwargs))), $(pullback_expr))
         end
-        function ChainRulesCore.rrule($(primal_sig_parts...))
+        function ($ChainRulesCore.rrule)($(esc_primal_sig_parts...))
             $(__source__)
-            return ($primal_invoke, $pullback_expr)
+            return ($(esc(primal_invoke)), $(pullback_expr))
         end
```

Comment:
Since this is now escaped correctly, there shouldn't be a need to interpolate something like `ChainRules` anymore. That was just my proposed quick fix, since we didn't handle escaping correctly before.